src/kOS.Safe/Execution/CPU.cs

```diff
@@ -1205,39 +1243,38 @@ private void ContinueExecution(bool doProfiling)
         {
             var executeNext = true;
             int howManyMainLine = 0;
-            bool reachedMainLine = false;
+            currentRunSection = Section.Trigger; // assume we begin with trigger mode until we hit mainline code.
+            
             executeLog.Remove(0, executeLog.Length); // In .net 2.0, StringBuilder had no Clear(), which is what this is simulating.
 
-            EndWait(); // Temporarily, perhaps. If the first thing it finds is the same OpcodeWait, it can revert back to wait mode.
-
-            while (currentStatus == Status.Running &&
-                   instructionsSoFarInUpdate < instructionsPerUpdate &&
+            while (instructionsSoFarInUpdate < instructionsPerUpdate &&
                    executeNext &&
                    currentContext != null)
             {
-                if (InstructionPointer == firstInstructionPointerInUpdate &&
-                    ! stack.HasTriggerContexts()) // This additional check is important.
-                                                 // Hypothetically it could be possible for trigger code to hit the same
-                                                 // IP as where mainline code was interrupted, if both trigger code and
-                                                 // mainline code make use of the same library function, and the
-                                                 // trigger interruption occurred when mainline code was in that
-                                                 // same library function.  This ensures that if so, that doesn't count.
-                                                 // It only counts when all the triggers have popped off the call stack.
+                if (! stack.HasTriggerContexts())
```

Comment:
With my new changes to ``stack.HasTriggerContexts()``, it's no longer a super expensive walk of the stack, so it's fine to just query it every time instead of trying to protect around it with the former messy logic.